### PR TITLE
Add table for timestamps

### PR DESCRIPTION
**Pull request by njacazio**
- Add object to convert run number to initial timestamp of run
- Add task to make timestamps and fill table
- Add consumer task for timestamp
- Add timestamp table to `AnalysisDataModel.h`

@ktf @jgrosseo @shahor02 @chiarazampolli I opened the PR :)

### CODE REVIEW

**Comment by njacazio** [comment]
Apologies for the long list of commits, also I added https://github.com/AliceO2Group/AliceO2/pull/3869/files#diff-348d7ed14f6b3b97ed63b87452235128 that is a python script to update the CCDB with objects in a non-timestamp-overlapping way, I don't know if this should go in this PR, apologies..

File: Analysis/Tasks/timestamp.cxx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    ccdb->setTimestamp(timestamp.value);
+    ccdb->setCachingEnabled(true);
+    // Not later than now objects
+    ccdb->setCreatedNotAfter(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count());
+  }
+
+  void process(soa::Join<aod::BCs, aod::Timestamps> const& iter)
+  {
+    for (auto i : iter) {
+      i.runNumber();
```
Comment (jgrosseo) [inline Analysis/Tasks/timestamp.cxx:70]: What is this doing?

File: Analysis/Tasks/timestamp.cxx [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  Configurable<std::string> url{"ccdb-url", "http://ccdb-test.cern.ch:8080", "url of the ccdb repository"};
+  Configurable<long> timestamp{"ccdb-timestamp", -1, "timestamp of the object"};
+
+  void init(o2::framework::InitContext&)
+  {
+    ccdb->setURL(url.value);
+    ccdb->setTimestamp(timestamp.value);
+    ccdb->setCachingEnabled(true);
+    // Not later than now objects
+    ccdb->setCreatedNotAfter(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count());
```
Comment (jgrosseo) [inline Analysis/Tasks/timestamp.cxx:64]: Can we add a parameter to the analysis task for this?

File: Analysis/Tasks/timestamp.cxx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  {
+    long timestamp = converter->getTimestamp(bc.runNumber());
+    InteractionRecord current(bc.globalBC(), 0);
+    InteractionRecord initial = o2::raw::HBFUtils::Instance().getFirstIR();
+    timestamp += 1000 * (current - initial).bc2ns();
+    ts_table(timestamp);
+  }
+};
+
+struct TimestampUserTask {
```
Comment (jgrosseo) [inline Analysis/Tasks/timestamp.cxx:52]: Let's factor this task out into a new file in the folder Tutorials and call it CCDBAccess (or similar)

File: Analysis/Tasks/update_ccdb.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/python3
```
Comment (jgrosseo) [inline Analysis/Tasks/update_ccdb.py:1]: Let's add a folder Analysis/scripts for this

File: Framework/Core/include/Framework/AnalysisDataModel.h [hunk 5]
```diff
@@ -35,6 +35,14 @@ DECLARE_SOA_TABLE(BCs, "AOD", "BC", o2::soa::Index<>,
                   bc::TriggerMask);
 using BC = BCs::iterator;
 
+namespace ts
+{
+DECLARE_SOA_INDEX_COLUMN(BC, bc);
```
Comment (jgrosseo) [inline Framework/Core/include/Framework/AnalysisDataModel.h:40]: The index can be removed. You use it with join later, so it is mapped 1-1 to the BC table (each BC entry has one corresponding Timestamp entry)

File: Framework/Core/include/Framework/AnalysisDataModel.h [hunk 6]
```diff
@@ -35,6 +35,14 @@ DECLARE_SOA_TABLE(BCs, "AOD", "BC", o2::soa::Index<>,
                   bc::TriggerMask);
 using BC = BCs::iterator;
 
+namespace ts
```
Comment (jgrosseo) [inline Framework/Core/include/Framework/AnalysisDataModel.h:38]: Please call the namespace timestamp

File: Analysis/Tasks/timestamp.cxx (hunk 1 above)
Comment (njacazio) [reply Analysis/Tasks/timestamp.cxx:70]: Nothing, vestige from the past ;)

File: Analysis/Tasks/update_ccdb.py (hunk 4 above)
Comment (njacazio) [reply Analysis/Tasks/update_ccdb.py:1]: Ok

File: Framework/Core/include/Framework/AnalysisDataModel.h (hunk 6 above)
Comment (njacazio) [reply Framework/Core/include/Framework/AnalysisDataModel.h:38]: Ok, good point

File: Framework/Core/include/Framework/AnalysisDataModel.h (hunk 5 above)
Comment (njacazio) [reply Framework/Core/include/Framework/AnalysisDataModel.h:40]: Ok, done

File: Analysis/Tasks/timestamp.cxx (hunk 2 above)
Comment (njacazio) [reply Analysis/Tasks/timestamp.cxx]: Done

File: Analysis/Tasks/update_ccdb.py (hunk 4 above)
Comment (njacazio) [reply Analysis/Tasks/update_ccdb.py]: Shall I add the CERN copyright statement?

File: Analysis/Tasks/timestamp.cxx (hunk 3 above)
Comment (njacazio) [reply Analysis/Tasks/timestamp.cxx]: Done

File: Analysis/Tasks/update_ccdb.py (hunk 4 above)
Comment (jgrosseo) [reply Analysis/Tasks/update_ccdb.py]: Yes

File: Analysis/Tasks/ccdbtest.cxx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    }
+  }
+};
+
+// This is how you can define your processing in a declarative way
+WorkflowSpec defineDataProcessing(ConfigContext const&)
+{
+  return WorkflowSpec{
+    adaptAnalysisTask<CCDBTask>("ccdbTestTask")};
+  // return WorkflowSpec{
```
Comment (jgrosseo) [inline Analysis/Tasks/ccdbtest.cxx:59]: Should these commented lines go?

File: Analysis/Tasks/update_ccdb.py (hunk 4 above)
Comment (ktf) [reply Analysis/Tasks/update_ccdb.py]: Also please make sure you do not use `/usr/bin/python3` but `/usr/bin/env python3`.

**Comment by jgrosseo** [comment]
@chiarazampolli Do you have comments on this PR?
Finally, the aim is to have something like this in a task: CCDBConfigurable<TH1> ccdbObj {"CCDBPath"} which is then always updated to the current timestamp

File: Analysis/Tasks/ccdbtest.cxx (hunk 7 above)
Comment (njacazio) [reply Analysis/Tasks/ccdbtest.cxx:59]: Maybe we can get rid of this task which is now superseded by the one in the tutorial

File: Analysis/Tasks/update_ccdb.py (hunk 4 above)
Comment (njacazio) [reply Analysis/Tasks/update_ccdb.py]: Thanks, it is now donw

File: Analysis/Tasks/ccdbtest.cxx (hunk 7 above)
Comment (njacazio) [reply Analysis/Tasks/ccdbtest.cxx]: Task terminated

**Comment by jgrosseo** [comment]
@njacazio Few things before it can be merged:
- Please rebase and fix the CMakeList conflict (came from another PR)
- Change long to uint64_t
I don't find anything wrong in the checkcode log. Either I overlook it, or there is a transient error.

**Comment by ktf** [comment]
The errors with the code checker are:

```
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Analysis/Core/include/Analysis/RunToTimestamp.h:27:3: error: annotate this function with 'override' or (rarely) 'final' [modernize-use-override]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Analysis/Core/include/Analysis/RunToTimestamp.h:27:3: error: annotate this function with 'override' or (rarely) 'final' [modernize-use-override]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Analysis/Core/include/Analysis/RunToTimestamp.h:27:3: error: annotate this function with 'override' or (rarely) 'final' [modernize-use-override]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Analysis/Core/include/Analysis/RunToTimestamp.h:27:3: error: annotate this function with 'override' or (rarely) 'final' [modernize-use-override]
```

use `final` rather than `virtual` if you do not want your method to be overridden further, use `override` if you do.

**Comment by njacazio** [comment]
> @njacazio Few things before it can be merged:
> 
>     * Please rebase and fix the CMakeList conflict (came from another PR)
> 
>     * Change long to uint64_t
>       I don't find anything wrong in the checkcode log. Either I overlook it, or there is a transient error.

@jgrosseo the choice for long was dictated by the ccdb api where the timestamp is a `long`:
```
  template <typename T>
  T* retrieveFromTFileAny(std::string const& path, std::map<std::string, std::string> const& metadata,
                          long timestamp = -1, std::map<std::string, std::string>* headers = nullptr, std::string const& etag = "",
                          const std::string& createdNotAfter = "", const std::string& createdNotBefore = "") const;
```
I can do the change but then wouldn't one need to change it also in the api and in the manager?

**Comment by chiarazampolli** [comment]
> @chiarazampolli Do you have comments on this PR?
> Finally, the aim is to have something like this in a task: CCDBConfigurable ccdbObj {"CCDBPath"} which is then always updated to the current timestamp

Hello @jgrosseo 

I am not familiar with "Configurable" objects, but in the code by Nicolo' they are the parameters to configure the CCDB (path, timestamp) not the object, like in your comment. Right? 

Also I have some questions, maybe coming from the fact that am not used to analysis tasks, but to devices. In devices, the run function is the one called at every timeframe, so here I would get the updated object. Same, I think, in the analysis tasks, but I have a doubt. In Analysis/Tutorials/src/ccdbaccess.cxx , you set the timestamp of the ccdb manager in the "init" function. Why? You don't use it later, because in fact in the "process" function you have a loop of timestamps built somehow (I don't know how, even if I understand that you build a joined table BC <--> timestamp). 

What I need is something like:

```
init() {
Service<o2::ccdb::BasicCCDBManager> ccdb;
// configure the ccdb with the path
...
}
process() {
// get somehow the timestamp
// get the object through the ccdb
}
```

When discussing with @ktf, I understood that for processing we would have not needed anything like the above, but we would have just depended on another workflow that would have provided as input to the processing one the updated CCDB objects in a transparent way. This is what the tof-dummy-ccdb workflow "pretends" to do (just it sends one object created ad hoc, without going to CCDB, and without checking if it needs to be updated).

Probably it is difficult to explain in git, it might be better to have a chat with you, @njacazio , @ktf  and possibly @costing and @shahor02 . Let me know.

Chiara

**Comment by jgrosseo** [comment]
Hi @chiarazampolli 

> I am not familiar with "Configurable" objects, but in the code by Nicolo' they are the parameters to configure the CCDB (path, timestamp) not the object, like in your comment. Right?

They can be used for simple parameters as Nicolo does, but here I really mean what I wrote (and which does not exist yet). You tell with some template that you need a certain CCDB object and that's it. The rest happens behind the scenes (for analysis it will update on each BC change, and for your case on each timestamp). You then just access the object.

> In Analysis/Tutorials/src/ccdbaccess.cxx , you set the timestamp of the ccdb manager in the "init" function. Why? 
I agree this is not needed. @njacazio can you comment?

**Comment by njacazio** [comment]
> > In Analysis/Tutorials/src/ccdbaccess.cxx , you set the timestamp of the ccdb manager in the "init" function. Why?
> 
> I agree this is not needed. @njacazio can you comment?
It was just to illustrate the features of the manager, I have now added an object retrieval test in the init function as well.
But of course, I can remove it if needed

**Comment by jgrosseo** [comment]
> 
> 
> > > In Analysis/Tutorials/src/ccdbaccess.cxx , you set the timestamp of the ccdb manager in the "init" function. Why?
> > 
> > 
> > I agree this is not needed. @njacazio can you comment?
> 
> It was just to illustrate the features of the manager, I have now added an object retrieval test in the init function as well.
> But of course, I can remove it if needed

It is not a big deal, but I think it may distract from the use case which we want to show here, and that how to get an object based on the timestamp from the BC. It can be confusing because the timestamp is set globally as a configurable parameter.

**Comment by njacazio** [comment]
> > > > In Analysis/Tutorials/src/ccdbaccess.cxx , you set the timestamp of the ccdb manager in the "init" function. Why?
> > > 
> > > 
> > > I agree this is not needed. @njacazio can you comment?
> > 
> > 
> > It was just to illustrate the features of the manager, I have now added an object retrieval test in the init function as well.
> > But of course, I can remove it if needed
> 
> It is not a big deal, but I think it may distract from the use case which we want to show here, and that how to get an object based on the timestamp from the BC. It can be confusing because the timestamp is set globally as a configurable parameter.

Ok I'm removing it then

**Comment by njacazio** [comment]
Hi, could we merge this?

**Comment by chiarazampolli** [comment]
For me it is fine. I still think I need to understand how to use this approach with data, as the plan with @ktf was different. 
I might call for a chat with the involved people.

Thanks!

Chiara

PS. @njacazio , will you implement also the "notBefore" feature for the timeMachine as discussed? Just to know the timescale, if you plan to do so, as it might be interesting also for calibration.

**Comment by ktf** [comment]
@chiarazampolli the plan is still the same. All this will be hidden longer term. What we can already easily do is to make sure that the Service<CCDBManager> is configured centrally, and not by the init method.

**Comment by ktf** [comment]
@njacazio could you squash your branch?

**Comment by njacazio** [comment]
> PS. @njacazio , will you implement also the "notBefore" feature for the timeMachine as discussed? Just to know the timescale, if you plan to do so, as it might be interesting also for calibration.

Ciao @chiarazampolli , this is already done

**Comment by chiarazampolli** [comment]
> > PS. @njacazio , will you implement also the "notBefore" feature for the timeMachine as discussed? Just to know the timescale, if you plan to do so, as it might be interesting also for calibration.
> 
> Ciao @chiarazampolli , this is already done

Wonderful, thanks!

**Pull request merged**